File: src/main/java/cat/kittens/mods/controller/input/ControllerInputProcessingThread.java
package cat.kittens.mods.controller.input;

import cat.kittens.mods.controller.ControllerSupport;
import cat.kittens.mods.controller.mixin.accessor.MinecraftAccessor;

public class ControllerInputProcessingThread extends Thread {
    private final ControllerAimingHandler aiming;

    public ControllerInputProcessingThread() {
        this.aiming = new ControllerAimingHandler();
        setName("Controller Input Processing");
        setDaemon(true);
    }

    @Override
    public void run() {
        var instance = MinecraftAccessor.instance();
        while (instance != null && instance.running)
            tick(((MinecraftAccessor) instance).timer().field_2370);
    }

    public void tickMappings() {
        ControllerSupport.support().manager().currentController().ifPresent(controller -> {
            var minecraft = MinecraftAccessor.instance();
            if (minecraft.currentScreen == null) {
                var mc = (MinecraftAccessor) minecraft;
                if (mc.lastInteraction() > mc.currentTicks())
                    mc.setLastInteraction(mc.currentTicks());
            }
            for (var action : MappingActions.registered()) {
                boolean isContextAllowed = false;
                for (var ctx : action.contexts()) {
                    if (MappingActionContext.current() == ctx) {
                        isContextAllowed = true;
                        break;
                    }
                }
                if (!isContextAllowed)
                    continue;
                ControllerSupport.support().mapping().find(action.id()).flatMap(mapping -> mapping.getContextFor(controller)).ifPresent(ctx -> {
                    ControllerSupport.support().setCurrentInputMethod(true);
                    if (!action.isExecutorInternalUseOnly())
                        action.executor().perform(ctx);
                });
            }
        });
    }

    public void tick(float delta) {
        ControllerSupport.support().manager().tick();
        aiming.tick(delta);
    }
}

File: src/main/java/cat/kittens/mods/controller/input/processing/ControllerInputProcessingThread.kt
package cat.kittens.mods.controller.input.processing

import cat.kittens.mods.controller.ControllerSupport
import cat.kittens.mods.controller.input.MappingActionContext
import cat.kittens.mods.controller.input.MappingActions
import cat.kittens.mods.controller.mixin.accessor.MinecraftAccessor
import kotlinx.coroutines.*
import net.minecraft.client.Minecraft
import org.lwjgl.input.Keyboard

private suspend inline fun CoroutineScope.withMinecraft(block: (mc: Minecraft) -> Unit) {
    while (isActive) {
        if (MinecraftAccessor.instance() != null)
            break
        delay(1L)
    }
    while (isActive) {
        block(MinecraftAccessor.instance())
    }
}

internal suspend fun inputProcessingCoroutineScope() = coroutineScope {
    val aiming = ControllerAimingHandler()
    launch(CoroutineName("Controller Input Processing")) {
        withMinecraft {
            ControllerSupport.manager.tick()
            tickMappings()
        }
    }
    launch(CoroutineName("Controller Aim Processing")) {
        withMinecraft {
            aiming.tick((it as MinecraftAccessor).timer().field_2370)
        }
    }
}

private fun tickMappings() {
    val controller = ControllerSupport.manager.currentController ?: return
    val minecraft = MinecraftAccessor.instance()
    if (minecraft.currentScreen == null) {
        val mc = minecraft as MinecraftAccessor
        if (mc.lastInteraction() > mc.currentTicks()) mc.setLastInteraction(mc.currentTicks())
    }
    ControllerSupport.isControllerActive = false
    for (action in MappingActions.all()) {
        if (action.contexts.none { a -> MappingActionContext.current == a })
            continue
        val ctx = ControllerSupport.mappingView[action.id]?.getContextFor(controller)
        if (ctx != null) {
            ControllerSupport.isControllerActive = true
            if (!action.isExecutorInternalUseOnly)
                action.executor.perform(ctx)
        }
    }
    if (Keyboard.getEventKeyState())
        ControllerSupport.isControllerActive = false
}

